### DISCUSSION TITLE
jukeboxControl's set parameter

### DISCUSSION THREAD

**Discussion by sentriz** [opensubsonic/open-subsonic-api#71] [General]
I'm trying to figure out jukeboxControl's `set` parameter, specifically what happens when the a playlist is "updated by overwriting"

for example say the current play queue is 

```
track 1
track 2 
track 3 <- current track
```

which was originally set with `?action=set&id=1&id=2&id=3` followed by an `?action=skip&index=2`

next, using dsub (presumably close to the original subsonic client for android) 
we append a track 4, which issues an `?action=set&id=1&id=2&id=3&id=4`

what should be the state of the player?

i noticed @spl0k, in supysonic, the state will be 

```
track 1 <- reported current track
track 2 
track 3 <- the one actually playing 
track 4
```

which is the same behavour as gonic right now, and it doesn't seem right

i originally had some logic that would restore the playing index after an operation like this, but that breaks shuffling seemingly 

and insight on this @spl0k or maybe @daneren2005? :grin: 

thanks!

**Comment by sentriz** [comment]
oh i think i've figured it out

- if the current playing track is not in the `set` list, reset everything and overwrite playlist to match list
- otherwise clear the playlist except what's currently playing, leaving the current track at index 0
- append the other `set` entries on top, in order
- if the currently playing track was not 0 in the incoming `set` list, restore the index to the original position

this seems to cover all the cases that DSub sends like

- play a playlist from index 0
- while being at a playlist index >0, add more tracks
- shuffle a playlist from index 0
- while being at a playlist index >0, shuffle
- while being at a playlist index >0, shuffle, add more tracks

**Comment by spl0k** [comment]
Hello 👋 

I think I actually followed the behaviour described by the [Subsonic API documentation](http://www.subsonic.org/pages/api.jsp#jukeboxControl) (or my interpretation of it) which states:

> `set` is similar to a `clear` followed by a `add`, but will not change the currently playing track.
